infrastructure/database/repo/users.py

```diff
@@ -16,31 +18,57 @@ async def get_or_create_user(
     ):
         """
         Creates or updates a new user in the database and returns the user object.
+        Only fields provided as parameters will be updated in the case of a conflict.
         :param user_id: The user's ID.
         :param full_name: The user's full name.
         :param language: The user's language.
         :param username: The user's username. It's an optional parameter.
         :return: User object, None if there was an error while making a transaction.
         """
+        values_dict = {
+            "user_id": user_id,
+            "username": username,
+            "full_name": full_name,
+            "language": language,
 
-        insert_stmt = (
-            insert(User)
-            .values(
-                user_id=user_id,
-                username=username,
-                full_name=full_name,
-                language=language,
-            )
-            .on_conflict_do_update(
-                index_elements=[User.user_id],
-                set_=dict(
+        }
+
+        # Remove None values to avoid nullifying columns in the database
+        values_dict = {
+            key: value for key, value in values_dict.items() if value is not None
+        }
+
+        try:
+
+            insert_stmt = (
+                insert(User)
+                .values(
+                    user_id=user_id,
                     username=username,
                     full_name=full_name,
-                ),
+                    language=language,
+                )
+                .on_conflict_do_update(
+                    index_elements=[User.user_id],
+                    set_={
+                        key: values_dict[key] for key in values_dict if key != "user_id"
+                    },
+                )
+                .returning(User)
             )
-            .returning(User)
-        )
-        result = await self.session.execute(insert_stmt)
+            result = await self.session.execute(insert_stmt)
 
-        await self.session.commit()
-        return result.scalar_one()
+            await self.session.commit()
+            return result.scalar_one()
+        except SQLAlchemyError as e:
+            await self.session.rollback()
```

Comment:
what kind of errors you're expecting here?

Reply:
Database Connection Issues, Data Integrity Errors, Transaction Errors, Timeouts and Resource Limits and etc